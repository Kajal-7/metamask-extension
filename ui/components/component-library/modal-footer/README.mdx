import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';

import { ModalFooter } from './modal-footer';

# ModalFooter

`ModalFooter` is a container component that handles the scrolling of the `Modal` content

<Canvas>
  <Story id="components-componentlibrary-modalfooter--default-story" />
</Canvas>

## Props

The `ModalFooter` accepts all props below as well as all [Box](/docs/components-ui-box--default-story#props) component props

<ArgsTable of={ModalFooter} />
