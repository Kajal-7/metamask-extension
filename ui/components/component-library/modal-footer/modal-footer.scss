.mm-modal-footer {
  /**
  *  TODO: Need to figure out the styles to allow for:
  *  - horizontally align children
  *  - wrap children if content overflows the width of the button
  *  - maintains a gap of 16px
  *  - children take up all available space whether horizontally or vertically aligned
  */
}
