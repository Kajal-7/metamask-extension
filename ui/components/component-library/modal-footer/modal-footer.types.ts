import React from 'react';

import type {
  StyleUtilityProps,
  PolymorphicComponentPropWithRef,
} from '../box';
import type { ButtonProps } from '../button';

export interface ModalFooterStyleUtilityProps extends StyleUtilityProps {
  /**
   * Additional className to add to the ModalFooter
   */
  className?: string;
  /**
   * The custom content of the ModalFooter
   */
  children?: React.ReactNode;
  /**
   * Array of buttons that will be displayed in the footer
   */
  buttonPropsArray?: ButtonProps<'button'>[];
  /**
   * The confirm button click event handler if this exists the confirm button will be displayed
   */
  onConfirm?: () => void;
  /**
   * Additional props to pass to the confirm button
   */
  confirmButtonProps?: ButtonProps<'button'>;
  /**
   * The cancel button click event handler if this exists the cancel button will be displayed
   */
  onCancel?: () => void;
  /**
   * Additional props to pass to the cancel button
   */
  cancelButtonProps?: ButtonProps<'button'>;
}

export type ModalFooterProps<C extends React.ElementType> =
  PolymorphicComponentPropWithRef<C, ModalFooterStyleUtilityProps>;

export type ModalFooterComponent = <C extends React.ElementType = 'span'>(
  props: ModalFooterProps<C>,
) => React.ReactElement | null;
