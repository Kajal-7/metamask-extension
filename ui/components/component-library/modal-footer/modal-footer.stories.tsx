import React from 'react';
import { StoryFn, Meta } from '@storybook/react';

import { ButtonVariant } from '..';
import { ModalFooter } from './modal-footer';

import README from './README.mdx';

export default {
  title: 'Components/ComponentLibrary/ModalFooter',
  component: ModalFooter,
  parameters: {
    docs: {
      page: README,
    },
  },
  argTypes: {
    className: {
      control: 'text',
    },
    children: {
      control: 'text',
    },
    buttonPropsArray: {
      control: 'array',
    },
    confirmButtonProps: {
      control: 'object',
    },
    onConfirm: {
      action: 'onConfirm',
    },
    cancelButtonProps: {
      control: 'object',
    },
    onCancel: {
      action: 'onCancel',
    },
  },
  args: {
    // buttonPropsArray: [
    //   { children: 'Confirm' },
    //   { children: 'Cancel', variant: ButtonVariant.Secondary },
    // ],
  },
} as Meta<typeof ModalFooter>;

const Template: StoryFn<typeof ModalFooter> = (args) => (
  <ModalFooter {...args} />
);

export const DefaultStory = Template.bind({});
DefaultStory.storyName = 'Default';
