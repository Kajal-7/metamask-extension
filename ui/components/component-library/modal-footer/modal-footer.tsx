import React from 'react';
import classnames from 'classnames';
import {
  BlockSize,
  Display,
  FlexWrap,
} from '../../../helpers/constants/design-system';
import { Box, Button, ButtonSize } from '..';
import type { PolymorphicRef, BoxProps } from '../box';
import type { ButtonProps } from '../button';
import { ButtonVariant } from '../button';

import { ModalFooterProps, ModalFooterComponent } from './modal-footer.types';

export const ModalFooter: ModalFooterComponent = React.forwardRef(
  <C extends React.ElementType = 'div'>(
    {
      className = '',
      children,
      buttonPropsArray,
      confirmButtonProps,
      onConfirm,
      cancelButtonProps,
      onCancel,
      ...props
    }: ModalFooterProps<C>,
    ref?: PolymorphicRef<C>,
  ) => (
    <Box
      className={classnames('mm-modal-footer', className)}
      ref={ref}
      paddingLeft={4}
      paddingRight={4}
      paddingBottom={4}
      paddingTop={2}
      display={Display.Grid}
      gap={4}
      // flexWrap={FlexWrap.Wrap}
      {...(props as BoxProps<C>)}
    >
      {children}
      {onCancel && (
        <Button
          onClick={onCancel}
          children="CancelCancelCancelCancelCancelCancelCancel"
          variant={ButtonVariant.Secondary}
          block
          {...(cancelButtonProps as ButtonProps<'button'>)}
          size={ButtonSize.Lg} // TODO: There is a type issue with using variant, size and spreading props after size
        />
      )}
      {onConfirm && (
        <Button
          size={ButtonSize.Lg}
          onClick={onConfirm}
          children="Confirm"
          block
          {...confirmButtonProps}
        />
      )}
      {buttonPropsArray &&
        buttonPropsArray.map(
          (buttonProp: ButtonProps<'button'>, index: number) => (
            <Button key={index} size={ButtonSize.Lg} {...buttonProp} />
          ),
        )}
    </Box>
  ),
);

export default ModalFooter;
