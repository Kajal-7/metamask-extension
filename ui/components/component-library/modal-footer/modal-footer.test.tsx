import React from 'react';
import { getByTestId, render } from '@testing-library/react';

import { ModalFooter } from './modal-footer';

describe('ModalFooter', () => {
  it('should render ModalFooter without error', () => {
    const { getByTestId } = render(<ModalFooter data-testid="modal-footer" />);
    expect(getByTestId('modal-footer')).toBeDefined();
    expect(getByTestId('modal-footer')).toHaveClass('mm-modal-footer');
  });
  it('should match snapshot', () => {
    const { container } = render(<ModalFooter />);
    expect(container).toMatchSnapshot();
  });
  it('should render with and additional className', () => {
    const { getByTestId } = render(
      <ModalFooter data-testid="modal-footer" className="test-class" />,
    );
    expect(getByTestId('modal-footer')).toHaveClass('test-class');
  });
  it('should fire the onConfirm function when clicked and pass additional props to the confirm button', () => {
    const onConfirm = jest.fn();
    const { getByText, getByTestId } = render(
      <ModalFooter
        onConfirm={onConfirm}
        confirmButtonProps={{ 'data-testid': 'confirm-button' }}
      />,
    );
    getByText('Confirm').click();
    expect(onConfirm).toHaveBeenCalled();
    expect(getByTestId('confirm-button')).toBeDefined();
  });
  it('should fire the onCancel function when clicked and pass additional props to the cancel button', () => {
    const onCancel = jest.fn();
    const { getByText, getByTestId } = render(
      <ModalFooter
        onCancel={onCancel}
        cancelButtonProps={{ 'data-testid': 'cancel-button' }}
      />,
    );
    getByText('Cancel').click();
    expect(onCancel).toHaveBeenCalled();
    expect(getByTestId('cancel-button')).toBeDefined();
  });
});
