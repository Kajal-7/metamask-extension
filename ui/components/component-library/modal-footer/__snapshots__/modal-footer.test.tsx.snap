// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalFooter should match snapshot 1`] = `
<div>
  <div
    class="mm-box mm-modal-footer mm-box--padding-top-2 mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--padding-left-4 mm-box--display-flex mm-box--gap-4"
  />
</div>
`;
